vms/platformvm/txs/mempool/mempool.go

```diff
@@ -4,42 +4,326 @@
+	minHeap            heap.Map[ids.ID, heapTx]
+	consumedUTXOs      *setmap.SetMap[ids.ID, ids.ID]
+	droppedTxIDs       *lru.Cache[ids.ID, error]
+	currentGas         gas.Gas
+	numTxsMetric       prometheus.Gauge
+	gasAvailableMetric prometheus.Gauge
 }
 
-func New(namespace string, registerer prometheus.Registerer) (*Mempool, error) {
-	metrics, err := txmempool.NewMetrics(namespace, registerer)
-	if err != nil {
-		return nil, err
+func New(
+	namespace string,
+	weights gas.Dimensions,
+	gasCapacity gas.Gas,
+	avaxAssetID ids.ID,
+	registerer prometheus.Registerer,
+) (*Mempool, error) {
+	numTxsMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "count",
+		Help:      "number of transactions in the mempool",
+	})
+
+	gasAvailableMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "gas_available",
+		Help:      "amount of remaining gas available",
+	})
+
+	if err := errors.Join(
+		registerer.Register(numTxsMetric),
+		registerer.Register(gasAvailableMetric),
+	); err != nil {
+		return nil, fmt.Errorf("failed to register metrics: %w", err)
+	}
+
+	m := &Mempool{
+		weights:     weights,
+		gasCapacity: gasCapacity,
+		avaxAssetID: avaxAssetID,
+		maxHeap: heap.NewMap[ids.ID, heapTx](func(a, b heapTx) bool {
+			return a.GasPrice > b.GasPrice
+		}),
+		minHeap: heap.NewMap[ids.ID, heapTx](func(a, b heapTx) bool {
+			return a.GasPrice < b.GasPrice
+		}),
+		consumedUTXOs:      setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs:       lru.NewCache[ids.ID, error](64),
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
 	}
-	pool := txmempool.New[*txs.Tx](
-		metrics,
-	)
-	return &Mempool{Mempool: pool}, nil
+
+	m.cond = lock.NewCond(&m.lock)
+	return m, nil
 }
 
 func (m *Mempool) Add(tx *txs.Tx) error {
-	switch tx.Unsigned.(type) {
-	case *txs.AdvanceTimeTx:
-		return ErrCantIssueAdvanceTimeTx
-	case *txs.RewardValidatorTx:
-		return ErrCantIssueRewardValidatorTx
-	default:
-		return m.Mempool.Add(tx)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return mempool.ErrConflictsWithOtherTx
+	}
+
+	ins, outs, producedAVAX, err := utxo.GetInputOutputs(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to get utxos: %w", err)
+	}
+
+	consumedAVAX := uint64(0)
+	for _, utxo := range ins {
+		if utxo.AssetID() != m.avaxAssetID {
+			continue
+		}
+
+		// The caller should verify txs but perform overflow checks anyway
+		consumedAVAX, err = math.Add(consumedAVAX, utxo.In.Amount())
+		if err != nil {
+			return fmt.Errorf("failed to add consumed AVAX: %w", err)
+		}
+	}
+
+	if consumedAVAX == 0 {
+		return errMissingConsumedAVAX
+	}
+
+	for _, utxo := range outs {
+		if utxo.AssetID() != m.avaxAssetID {
+			continue
+		}
+
+		// The caller should verify txs but perform overflow checks anyway
+		producedAVAX, err = math.Add64(producedAVAX, utxo.Out.Amount())
```

Comment:
`Add64` is deprecated.
```suggestion
		producedAVAX, err = math.Add(producedAVAX, utxo.Out.Amount())
```